rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,706 @@
+    OnGraphServerStop(ctx *core.GraphServerContext) error
+}
+```
+
+## Example Use Cases
+
+- **Advanced GraphQL Operation Handling**: A module that walks the parsed operation and performs custom logic based on the operation type, fields, and arguments.
+- **Request validation**: A module that validates incoming requests and returns an error if the request is invalid.
+- **Custom Telemetry**: A module that creates custom spans or metric data and sends it to a telemetry backend.
+- **Custom Authentication / Authorization**: A module that authenticates incoming requests and adds user information to the subgraph requests.
+- **Response interception**: A module that intercepts responses from subgraphs and modifies them before they are sent to the client.
+- **Response Caching**: A module that caches responses from subgraphs and returns them for identical requests.
+- **GraphQL Directive Handling**: A module that reacts to specific GraphQL directives in the operation definition.
+- **GraphQL Scalar Handling**: A module that allows custom handling of scalar types in the operation e.g. validation, transformation.
+- **Enriching Logs**: A module that adds custom log fields to the router and subgraph logs.
+
+## Backwards Compatibility
+
+The new module system is not backwards compatible with the old module system. Existing custom modules will need to be rewritten to use the new interfaces and API. We will provide a migration guide and tooling to help developers migrate their custom modules to the new system.
+
+# Example Modules
+
+__All examples are pseudocode and not tested, but they are as close as possible to the final implementation__
+
+## Custom Telemetry
+
+This module adds custom attributes to the OpenTelemetry span for each router request. Data can come from the request, the router configuration, or external sources. The example modifies the first span of the router but depending on the hook, the span is different.
+
+```go
+type MyModule struct{}
+
+// Ensure that MyModule implements the RouterRequestHook interface
+var _ RouterRequestHook = (*MyModule)(nil)
+
+func (m *MyModule) OnRouterRequest(req *core.RouterRequest, err error) error {
+	req.Telemetry.Span.AddEvent("Router Request")
+	req.Telemetry.Span.AddAttributes(
+		attribute.String("customer.id", "123"),
+	)
+	return nil
+}
+```
+
+## Intercepting Router Responses
+
+This module intercepts the final Router response to rewrite errors and add custom extensions. You have access to the parsed operation, the router request, and the router response.
+
+```go
+type MyModule struct{}
+
+// Ensure that MyModule implements the RouterResponseHook interface
+var _ RouterResponseHook = (*MyModule)(nil)
+
+func (m *MyModule) OnRouterResponse(res *core.RouterResponse, err error) error {
+	// Add custom extensions to the response
+	res.Response.Extensions["myExtension"] = "myValue"
+
+	// Rewrite errors in the response
+	if len(res.Response.Errors) > 0 {
+		for _, err := range res.Response.Errors {
```

Comment:
1.1 We are interested in changing the message, as I understand we will now have this access.
1.2 It is not clear how we will work with untyped extensions, through fastjson? It's important for us not to change custom extensions that we didn't expect.
2. We would also like to understand which subgraph the error belongs to, so that we can change it with this logic in mind.
3. We want to do ctx.Set, so that we can get the modified errors in other methods (for example, for logging).